elements/rh-code-block/rh-code-block.ts

```diff
@@ -26,14 +45,162 @@ export class RhCodeBlock extends LitElement {
+               fill="currentColor"
+               viewBox="0 0 11 7">
+            <path d="M4.919.239.242 4.847a.801.801 0 0 0 0 1.148l.778.766a.83.83 0 0 0 1.165 0L5.5 3.495 8.815 6.76a.83.83 0 0 0 1.165 0l.778-.766a.802.802 0 0 0 0-1.148L6.08.239a.826.826 0 0 0-1.162 0Z"/>
+          </svg>
+        </button>
+      </div>
     `;
   }
+
+  protected override firstUpdated(): void {
+    this.#computeLineNumbers();
+  }
+
+  protected override updated(changed: PropertyValues<this>): void {
+    if (changed.has('wrap')) {
+      this.#computeLineNumbers();
+      for (const action of this.querySelectorAll('rh-code-action')) {
+        if (action.action === 'wrap') {
+          action.active = this.wrap;
+        }
+      }
+    }
+  }
+
+  /**
+   * Clone the text content and connect it to the document, in order to calculate the number of lines
+   * @license portions copyright prism.js authors (MIT license)
+   */
+  #computeLineNumbers() {
+    const slot = this.shadowRoot?.getElementById('content') as HTMLSlotElement;
+
+    const codes: HTMLElement[] = slot.assignedElements().flatMap(x =>
+        x instanceof HTMLScriptElement ||
+        x instanceof HTMLPreElement ? [x]
+      : []);
+
+    const infos: CodeLineHeightsInfo[] = codes.map(element => {
+      const sizer = document.createElement('span');
+      sizer.className = 'sizer';
+      sizer.innerText = '0';
+      sizer.style.display = 'block';
+      this.shadowRoot?.getElementById('sizers')?.appendChild(sizer);
+      return {
+        lines: element.textContent?.split(/\n(?!$)/g) ?? [],
+        lineHeights: [],
+        sizer,
+        oneLinerHeight: sizer.getBoundingClientRect().height,
+      };
+    });
+
+    for (const { lines, lineHeights, sizer, oneLinerHeight } of infos) {
+      lineHeights[lines.length - 1] = undefined; // why?
+      lines.forEach((line, i) => {
+        if (line.length > 1) {
+          const e = sizer.appendChild(document.createElement('span'));
+          e.style.display = 'block';
+          e.textContent = line;
+        } else {
+          lineHeights[i] = oneLinerHeight;
+        }
+      });
+    }
+
+    for (const { sizer, lineHeights } of infos) {
+      let childIndex = 0;
+      for (let i = 0; i < lineHeights.length; i++) {
+        if (lineHeights[i] === undefined) {
+          lineHeights[i] = sizer.children[childIndex++].getBoundingClientRect()?.height ?? 0;
+        }
+      }
+      sizer.remove();
+    }
+
+    this.#lineHeights = infos.flatMap(x =>
+      x.lineHeights?.map(y =>
+        `${y ?? x.oneLinerHeight}px` as const));
+
+    this.requestUpdate('#linesNumbers', 0);
+  }
+
+  #onCodeAction(event: RhCodeActionEvent) {
+    switch (event.action) {
+      case 'copy':
+        return this.#copy();
+      case 'wrap':
+        this.wrap = !this.wrap;
+        this.requestUpdate();
+        return;
+    }
+  }
+
+  #onClickExpand() {
+    this.fullHeight = !this.fullHeight;
+  }
+
+  async #copy() {
```

Comment:
There is no user feedback when copy is clicked, so it's hard to tell that code was copied.